ext/openssl/ossl_ssl.c

```diff
@@ -1322,15 +1322,80 @@ ossl_sslctx_add_certificate(int argc, VALUE *argv, VALUE self)
     return self;
 }
 
+/*
+ * call-seq:
+ *    ctx.add_certificate_chain_file(certs_path, pkey_path) -> self
+ *
+ * Loads chain certificates from _certs_path_ and a private key from
+ * _pkey_path_.
+ *
+ * === Parameters
+ * _certs_path_::
+ *   A path to a chain certificates file. It may be a single certificate.
+ *   An instance of String.
+ * _pkey_path_::
+ *   A path to a private key file. An instance of String.
+ *
+ * === Example
+ *   ctx.add_certificate_chain(rsa_cert_path, rsa_key_path)
+ *
+ *   ctx.add_certificate_chain(ecdsa_cert_path, ecdsa_key_path)
+ *
+ * === Note
+ * The file format of the certificate and private key must be PEM.
+ *
+ * The certificate file must be starting with the subject's certificate and
+ * followed by intermediate CA certificate(s).
+ *
+ * OpenSSL before the version 1.0.2 could handle only one extra chain across
+ * all key types. Calling this method discards the chain set previously.
+ */
 static VALUE
-ossl_sslctx_add_certificate_chain_file(VALUE self, VALUE path)
+ossl_sslctx_add_certificate_chain_file(VALUE self, VALUE certs_path, VALUE pkey_path)
 {
-    StringValue(path);
-    SSL_CTX *ctx = NULL;
+    SSL_CTX *ctx;
+    X509 *x509;
+    char *ccerts_path, *cpkey_path;
+    FILE *fp;
+    EVP_PKEY *pkey, *pub_pkey;
 
     GetSSLCTX(self, ctx);
 
-    return SSL_CTX_use_certificate_chain_file(ctx, RSTRING_PTR(path)) == 1 ? Qtrue : Qfalse;
+    /* Retrieve private key */
+    cpkey_path = StringValueCStr(pkey_path);
+    fp = fopen(cpkey_path, "r");
+    if (!fp)
+        rb_raise(rb_eArgError, "failed to open pkey file");
+    pkey = PEM_read_PrivateKey(fp, NULL, 0, NULL);
+    fclose(fp);
+    if (!pkey)
+        rb_raise(rb_eArgError, "failed to open pkey file");
+    /* Retrieve public key */
+    ccerts_path = StringValueCStr(certs_path);
+    fp = fopen(ccerts_path, "r");
+    if (!fp)
+        rb_raise(rb_eArgError, "failed to open certs file");
+    x509 = PEM_read_X509(fp, NULL, 0, NULL);
+    fclose(fp);
+    if (!x509)
+        rb_raise(rb_eArgError, "failed to open certs file");
+    pub_pkey = X509_get_pubkey(x509);
+    /* The reference counter is bumped, and decremented immediately. */
+    EVP_PKEY_free(pub_pkey);
+    if (!pub_pkey)
+        rb_raise(rb_eArgError, "certificate does not contain public key");
+
+    if (EVP_PKEY_cmp(pub_pkey, pkey) != 1)
```

Comment:
I need a review.

Using `SSL_CTX_check_private_key`,  `ossl_sslctx_add_certificate_chain_file` does not needs to call fopen.
To check arguments in advance, I chose not to use it.
It should use `SSL_CTX_check_private_key`?